services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -403,14 +416,24 @@ private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel re
         TokenValidation validation) throws OAuthErrorException {
         if (realm.isRevokeRefreshToken()) {
             AuthenticatedClientSessionModel clientSession = validation.clientSessionCtx.getClientSession();
-            if (validateTokenReuse(session, realm, refreshToken, clientSession, true)) {
+            try {
+                validateTokenReuse(session, realm, refreshToken, clientSession, true);
                 int currentCount = clientSession.getCurrentRefreshTokenUseCount();
                 clientSession.setCurrentRefreshTokenUseCount(currentCount + 1);
+            } catch (OAuthErrorException oee) {
+                if (logger.isDebugEnabled()) {
+                    logger.debugf("Failed validation of refresh token %s due it was used before. Realm: %s, client: %s, user: %s, user session: %s. Will detach client session from user session",
+                            refreshToken.getId(), realm.getName(), clientSession.getClient().getClientId(), clientSession.getUserSession().getUser().getUsername(), clientSession.getUserSession().getId());
+                }
+                clientSession.detachFromUserSession();
+                new ResourceAdminManager(session).logoutClientSession(realm, clientSession.getClient(), clientSession);
```

Comment:
Do we want to eventually send logout events to clients? Perhaps just detach the client session is enough?

Reply:
Question is why not to send logout event? Any issue with that?

Usually when some possible attack is detected, we send an event to client (EG. when admin or user updates "notBefore" . Or when admin or user logouts through admin console or account console). The event is sent just in the case when client is using Keycloak server-side adapter and has "admin URL" configured (proprietary Keycloak mechanism).

Reply:
You pretty much touched my point. We only send under specific circumstances. Make this possible also during validation is something I would avoid.

Depending on how many clients and how they are configured, you have an additional cost. With no real benefit as we have other means to advertise clients to invalidate tokens/sessions.

Reply:
I am not convinced either way, but I am just removing it from this PR. From what I know, we might be more align to remove Keycloak proprietary "OIDC extensions" in the future anyway.

Reply:
Ok :) I keep the "started" due the compatibility with UserSessionModel.